### PR TITLE
STEP 2 Activemodel7 couchbase ruby client

### PR DESCRIPTION
**Pull request by DamienVoreiter**

### CODE REVIEW

File: lib/couchbase-orm/connection.rb [hunk 1]
```diff
@@ -1,3 +1,4 @@
+
```
Comment (DamienVoreiter) [inline lib/couchbase-orm/connection.rb:1]:
```suggestion
```

File: lib/couchbase-orm/base.rb [hunk 2]
```diff
@@ -96,6 +96,14 @@ def attribute(*names, **options)
                 end
             end
 
+            def column_names
+                @attributes.map { |col| col[0].to_s }
+            end
+
+            def attribute_names
+                @attributes.map { |col| col[0].to_s }
+            end
```
Comment (DamienVoreiter) [inline lib/couchbase-orm/base.rb:105]:
```suggestion
alias :column_names :attribute_names
```

File: lib/couchbase-orm/base.rb [hunk 3]
```diff
@@ -199,7 +207,10 @@ def id=(value)
         end
 
         def read_attribute(attr_name)
-            @__attributes__[attr_name]
+            read_fn = self.class.attributes[attr_name][:read_fn]
```
Comment (DamienVoreiter) [inline lib/couchbase-orm/base.rb:210]: useless

File: lib/couchbase-orm/n1ql.rb [hunk 4]
```diff
@@ -35,7 +35,7 @@ module ClassMethods
             #    # ...
             #  end
             # TODO: add range keys [:startkey, :endkey]
-            def n1ql(name, query_fn: nil, emit_key: [], **options)
+            def n1ql(name, select: nil, query_fn: nil, emit_key: [], custom_order: nil, **options)
```
Comment (DamienVoreiter) [inline lib/couchbase-orm/n1ql.rb:38]: TODO remove select

File: lib/couchbase-orm/n1ql.rb [hunk 5]
```diff
@@ -96,8 +98,8 @@ def convert_values(values)
 
             def build_where(keys, values)
                 where = keys.each_with_index
-                            .reject { |key, i| values.try(:[], i).nil? }
-                            .map { |key, i| "#{key} = #{values[i] }" }
+                            .select { |_key, i| i < values.count }
```
Comment (DamienVoreiter) [inline lib/couchbase-orm/n1ql.rb:101]: change this

File: .gitignore [hunk 6]
```diff
@@ -1,4 +1,5 @@
 *.sw?
+.rbenv-vars
```
Comment (giallon) [inline .gitignore:2]: already set at line 4

File: spec/base_spec.rb [hunk 7]
```diff
@@ -2,169 +2,170 @@
...
+  it "should support serialisation" do
+    base = BaseTest.create!(name: 'joe')
 
-        base.destroy
-    end
+    base_id = base.id
+    expect(base.to_json).to eq({ id: base_id, name: 'joe', job: nil, prescribing_date:nil }.to_json)
+    expect(base.to_json(only: :name)).to eq({ name: 'joe' }.to_json)
 
-    it "should support dirty attributes" do
```
Comment (giallon) [inline spec/base_spec.rb:75]: why delete test 'support dirty attributes' ?

File: .gitignore (hunk 6 above)
Comment (DamienVoreiter) [inline .gitignore:2]:
```suggestion
```

File: spec/base_spec.rb (hunk 7 above)
Comment (DamienVoreiter) [reply spec/base_spec.rb:75]: not deleted, just line 76 on the new file

File: spec/n1ql_spec.rb [hunk 8]
```diff
@@ -62,6 +69,33 @@ class N1QLTest < CouchbaseOrm::Base
         }
 
         expect(Set.new(docs)).to eq(Set.new(%w[bob jane]))
+
```
Comment (simkim) [inline spec/n1ql_spec.rb:72]: maybe should be split in another it bloc, target good practice is one expect per it

File: spec/base_spec.rb [hunk 9]
```diff
@@ -2,169 +2,170 @@
...
-        same_base = BaseTest.find(base.id)
-        expect(base).to eq(same_base)
-        expect(base).not_to be(same_base)
-        expect(base2).not_to eq(same_base)
-
-        base.delete
-        base2.delete
-        base3.delete
-    end
+  it "should be comparable to other objects" do
```
Comment (simkim) [inline spec/base_spec.rb:33]: maybe it's me but seems' like most expect have been removed rending the test useless (but it's needed)

**Review by simkim — CHANGES_REQUESTED** [review]

**Review by giallon — APPROVED** [review]

**Pull request merged**